### PR TITLE
Update props needed for Blank Connection Viewer

### PR DESCRIPTION
**Pull request by ben-polinsky**

### CODE REVIEW

File: packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx
Comment (calebmshafer) [inline packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx:99]: should this include the blankConnection as well? Or should we try and remove the need for the blankConnection as well?

Comment (ben-polinsky) [reply packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx:99]: Considered it, but decided to play it safe and allow an override for now.

Comment (ben-polinsky) [reply packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx:99]: Could extract the overrides to separate/different props...

Comment (aruniverse) [reply packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx:99]: actually doesnt really make sense for check the blankConnection because its optional now

File: packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx
Comment (aruniverse) [inline packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx:210]: if we're going to go with the approach where you only need an iTwinId then we should pass it as another arg here,
https://www.itwinjs.org/reference/core-frontend/imodelconnection/blankconnectionprops/

Comment (ben-polinsky) [reply packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx:210]: Yes, makes sense. Good catch.

Comment (aruniverse) [reply packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx:210]: idk if the arg should be required, because what if the user provides blankConnection instead?
i guess it depends on what we do with that prop, and if we can remove it w/o a breaking change

Comment (ben-polinsky) [reply packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx:210]: Was working under the assumption that iTwinId was always required, but I see now.

Comment (ben-polinsky) [reply packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx:210]: I can just make the blankConnection required again, if that is easiest

File: packages/modules/viewer-react/src/tests/services/iModelViewerHelper.test.ts
Comment (ben-polinsky) [inline packages/modules/viewer-react/src/tests/services/iModelViewerHelper.test.ts:20]: I really wanted to extract this and use it in the BaseInitializer test, but had issues with jest hoisting/initializing modules.

File: packages/modules/viewer-react/src/types.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/types.ts:144]: i think we should leave this as required, as we need both itwin and imodel ids to open a remote connection model

File: packages/modules/viewer-react/src/types.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/types.ts:152]: wondering if we should split this type further down and have it be a union of the deprecated way and the updated/recommended usage. youre kina already doing this 

as is rn, BlankViewerProps === BlankViewerPropsWithoutDeprecation (ignoring the deprecated jsdoc tag)

File: packages/modules/viewer-react/src/types.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/types.ts:172]: is it really _required_ if everything is optionally defined?

File: packages/modules/viewer-react/src/types.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/types.ts:182]: this really shouldnt be a valid case, and i think the way the types were setup if you didnt satisfy the XOR union it would yell at you that youre missing props

File: packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts:33]: i was on a TS craze when i did my refactor for 3.0; wondering if we should use type guards instead

File: packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts:58]: we should avoid using non null assertions imo, i know everything is defined here because of connectionType, but i think we can get TS to correctly infer the types if we use type guards

File: packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts
Comment (ben-polinsky) [reply packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts:33]: Yep, I can clean this up.

File: packages/modules/viewer-react/src/types.ts
Comment (ben-polinsky) [reply packages/modules/viewer-react/src/types.ts]: Removed ConnectionType. Was adding it to aid in discriminated unions, but using your `XOR` was fine.

File: packages/modules/viewer-react/src/types.ts
Comment (ben-polinsky) [reply packages/modules/viewer-react/src/types.ts]: Revamped.

File: packages/modules/viewer-react/src/types.ts
Comment (ben-polinsky) [reply packages/modules/viewer-react/src/types.ts]: Also reworked.

File: packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx
Comment (ben-polinsky) [reply packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx]: Reworked.

File: packages/modules/viewer-react/src/types.ts
Comment (ben-polinsky) [reply packages/modules/viewer-react/src/types.ts]: Done.

File: packages/modules/viewer-react/src/types.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/types.ts:187]: i think we can delete this, especially as we are able to infer the type

File: packages/modules/viewer-react/src/types.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/types.ts:152]: i know GuidString is just an alias for string, but i guess might as well change iModelId as well if we're changing iTwinId

File: packages/modules/viewer-react/src/types.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/types.ts:163]: thoughts on breaking these into their own types, so we can mark the whole subtype as deprecated, and not just the blankConnection property?

File: packages/modules/viewer-react/src/types.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/types.ts:168]: wonder if we need this one at all; it might be an unecessary step in between the deprecated api and where we're pushing the api fwd to; might as well have consumers just update to the preferred api.

File: packages/modules/viewer-react/src/types.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/types.ts:164]: i like this, but does this make it a breaking change?

File: packages/modules/viewer-react/src/types.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/types.ts:171]: if we keep this, should this be optional?

File: packages/modules/viewer-react/src/types.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/types.ts:144]: maybe we should relook at whats being exported; not a big deal; but everythign here is re-esported at the top level pkgs, and wonder if we should be more aware of what our public api/types are

File: packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts:47]: opening

File: packages/modules/viewer-react/src/components/BaseViewer.tsx
Comment (aruniverse) [inline packages/modules/viewer-react/src/components/BaseViewer.tsx:36]: surprised no ts errors here, nice

**Review by aruniverse — COMMENTED** [review]
looking good, i need to  test it out locally. the clever use of typing is helpful (imo), but sometimes the warnings in the IDE and error-overlay can get ugly, so just want to check that.

we will need to update our docs (the readmes) before this pr goes through with updated blank viewer prop changes

File: packages/modules/viewer-react/src/types.ts
Comment (ben-polinsky) [reply packages/modules/viewer-react/src/types.ts:152]: Yeah... I think I'll go back to string.

File: packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts
Comment (ben-polinsky) [inline packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts:72]: @aruniverse Now relying on your `ModelLoaderProps` as Viewer input, but narrowing down to required props here. My last attempt at creating some tighter types. If you're not a fan, we can revisit when `blankConnectionProps` are removed.

File: packages/modules/viewer-react/src/types.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/types.ts:146]: I think instead of doing 
```ts
Partial<ConnectedViewerProps & FileViewerProps & BlankViewerProps> & ...
```
using a union like this is more powerful:
```ts
(ConnectedViewerProps | FileViewerProps | BlankViewerProps) & ...
```

TS is able to better infer the union type, and the intellisense you get is much stronger imo
[here's an exmaple](https://www.typescriptlang.org/play?#code/C4TwDgpgBAwg9gOwRAxsCATAalAvFAbwFgAoKKAS2AHcKEAuKAZ2ACc6BzAblPIoFs4GCABtGLdgg6kAvqVKhIUAGIUREHPmJkoAMzUQACgENgAC3FtOs+SUXQAQiOMIA1psK8oAI2dvGBHIkQQrg0PD8ggh4njrGjAgArvzeEKwANF7ejN5wcOouNiShSgDyIhgxABTwSKjo2FAAPioGOC1OLu4AlFAAZLBwkYi29lAAchDUMSaswBTGIgA8tchomDgDquqbUJ1uWAB8-YPDCLYoiCxQcIzllVpe8VAAjJk62VBsiRDvfDR0RgAIiB7yKpEuCGuDAmUxi2nIzzeWUY31+XiotBhILBwSAA)

File: packages/modules/viewer-react/src/types.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/types.ts:151]: not the exact type from before, but its narrower so its a :thumbsup:

File: packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts:1]: nit: for consistency, could you export this file in the barrel in this dir? and then consume exports from here from that barrel in the other files?

File: packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx
Comment (aruniverse) [inline packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx:62]: nit: rename to requiredProps or requiredConnectionProps or something?

File: packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts
Comment (aruniverse) [reply packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts:72]: i like this a lot more

File: packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts:101]: should we return undefined instead? does that make more sense than false?

File: common/scripts/.eslintrc.ts.json
Comment (aruniverse) [inline common/scripts/.eslintrc.ts.json:11]: lets revert this

File: packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts:62]: once we revert the eslint change, we'll need to explicitly disable deprecation for this line, and the others, or maybe just the entire file

File: packages/modules/viewer-react/src/types.ts
Comment (aruniverse) [inline packages/modules/viewer-react/src/types.ts:174]: should we just move this into iModelViewHelper since no one else uses this?

File: common/changes/@itwin/viewer-react/benp-optional-imodel-id_2022-09-06-18-35.json
Comment (aruniverse) [inline common/changes/@itwin/viewer-react/benp-optional-imodel-id_2022-09-06-18-35.json:6]: maybe make this a minor bump?
and change the comment so its says something like blank connection will be created if iTwinId is supplied and iModelId is not? might need to work on the wording there

Comment (aruniverse) [reply common/changes/@itwin/viewer-react/benp-optional-imodel-id_2022-09-06-18-35.json:6]: might want to update the other changelog entry as well

**Review by aruniverse — COMMENTED** [review]
looking good, didnt review test code (i trust you there), havent ran locally yet

File: packages/modules/viewer-react/src/types.ts
Comment (ben-polinsky) [reply packages/modules/viewer-react/src/types.ts:146]: Yes, thanks makes sense.

Comment (ben-polinsky) [reply packages/modules/viewer-react/src/types.ts:146]: Oh the issue is that we need the possibility of accepting any of the props, and the union restricts the props to one of the types:

```ts
 export type ModelLoaderProps = (
  ConnectedViewerProps | FileViewerProps | BlankViewerProps
 ) & LoaderProps;
```
image

Comment (aruniverse) [reply packages/modules/viewer-react/src/types.ts:146]: so we cant destructure completely but would have to do props.iTwinId && props.iModelId or props.filePath instead?

Idk maybe not worth it, ill leave it up to you to resolve. thx for calling this out

File: packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts
Comment (ben-polinsky) [reply packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts:101]: Yep - will do. Leftover from when this was `hasRequiredConnectionProps`

File: packages/modules/viewer-react/src/types.ts
Comment (ben-polinsky) [reply packages/modules/viewer-react/src/types.ts:146]: Even worse, will have to do `if ("filePath" in props)...`

Comment (ben-polinsky) [reply packages/modules/viewer-react/src/types.ts:146]: I think we can circle back with some better types once we make some breaking changes

Comment (aruniverse) [reply packages/modules/viewer-react/src/types.ts:146]: yea that sounds ok to me

File: common/scripts/.eslintrc.ts.json
Comment (ben-polinsky) [reply common/scripts/.eslintrc.ts.json:11]: Figured this out.. for some reason cannot ignore the deprecation errors if I leave the deprecation/deprecation rule here. If I move it out to the base eslint config, then it works 🤷‍♂️

File: packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts
Comment (ben-polinsky) [reply packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts:65]: Done.

File: packages/modules/viewer-react/src/types.ts
Comment (ben-polinsky) [reply packages/modules/viewer-react/src/types.ts]: Done

File: packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts
Comment (ben-polinsky) [reply packages/modules/viewer-react/src/services/iModel/iModelViewerHelper.ts:1]: done

File: packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx
Comment (ben-polinsky) [reply packages/modules/viewer-react/src/components/iModel/IModelLoader.tsx]: done

File: common/changes/@itwin/viewer-react/benp-optional-imodel-id_2022-09-06-18-35.json
Comment (ben-polinsky) [reply common/changes/@itwin/viewer-react/benp-optional-imodel-id_2022-09-06-18-35.json:6]: Holding this open until I test changes in the test-app. Will add change then.

Comment (ben-polinsky) [reply common/changes/@itwin/viewer-react/benp-optional-imodel-id_2022-09-06-18-35.json:6]: Alright tests out fine. If you have time, take another pass at my responses/corrections and LGTM (let's get this merged :))

File: CONTRIBUTING.MD
Comment (aruniverse) [inline CONTRIBUTING.MD:63]:
```suggestion
4. Test changes with the appropriate development app for your changes (web-viewer-test, desktop-viewer-test).
```

File: common/scripts/.eslintrc.ts.json
Comment (aruniverse) [reply common/scripts/.eslintrc.ts.json]: might be due to multiple version of typescript-eslint in the repo now due to using cra5 for the test-apps which pulls in ts-eslint@5, and eslint@8. i might try to fix this in a different pr

**Review by aruniverse — APPROVED** [review]

File: common/changes/@itwin/viewer-react/benp-optional-imodel-id_2022-09-06-18-35.json
Comment (aruniverse) [reply common/changes/@itwin/viewer-react/benp-optional-imodel-id_2022-09-06-18-35.json]: lgtm!

**Pull request merged**
